accepted/cranelift-isel-isle-peepmatic.md

```diff
@@ -0,0 +1,926 @@
+  semantically undefined (hence flexible).
+- Single-pass lowering semantics via unrolling.
+
+We will dive into each of these principles in more detail below.
+
+# Intro: Term-Rewriting System
+
+The basic concept behind the sort of system that this DSL embodies is
+that of *term-rewriting*. In brief, this means that the input and
+output of the system are trees (or more general graphs) of *terms*
+(symbolic nodes with arguments), and the system transforms input to
+output by applying a set of *rewrite rules*. Each rule matches on some
+*pattern* -- for example, a term of a particular type with an input
+that matches some sub-pattern -- and replaces it with a new tree of
+terms.
+
+This is a natural abstraction for transforming code. If we define a
+term view of the IR on the input side and machine instructions on the
+output side, such that a term represents an operator that consumes
+values and produces a value, then the rewrite rules just define
+*equivalences*. These rules are easier to validate than other sorts of
+compiler-backend implementations because "all" one has to do is to
+show that the two sides of the equivalence are equal, separately for
+each rule (i.e., the validation is modular).
+
+An example of a rewrite rule might be:
+
+```plain
+
+    (rule (Iadd a b)
+          (X86Add a b))
+```
+
+which means that an `Iadd` "term" with two arguments `a` and `b` (the
+"left-hand side" or "pattern" of the rule) can be written to an
+`X86Add` term with the same arguments (the "right-hand side" or
+"expression"). One could imagine many such patterns for particular
+combinations that the target machine supports.
+
+The concept of a rewrite rule, transforming one term into another
+equivalent one, provides a *simple, understandable semantics*. The
+importance of this cannot be overstated: if the core of the DSL is
+complex because its design is intertwined with a particular backend
+design or strategy, or other incidental details, then one will not be
+able to use it effectively without understanding implementation
+details. On the other hand, if we have a clean rewrite-based
+semantics, one can easily convince oneself that a new rule is safe to
+add as long as it denotes a true equivalence. This is similar to how
+"equational reasoning", in functional programming terms (i.e., that
+one can always substitute a function invocation with its definition),
+makes a program's semantics easier to understand.
+
+The order in which rewrite rules will be applied can be influenced by
+user-specified priorities. However, the language definition is careful
+to *not* specify which rule must be applied first; priorities are just
+heuristics, or hints. This leaves room for the DSL use-case to apply
+rules differently: for example, we might test/fuzz a backend by
+applying rules in a different order. One should also be free to tweak
+priorities to attempt to optimize, without fear of breaking the
+semantics.
```

Comment:
It might make sense to tell when multiple rules can apply to the same input and maybe allow hints to acknowledge the overlap. These hints could then also contain the advisory priority or the rules.

Reply:
Yes, this gives me an idea for a DSL-compiler mode that produces some "explainer" output -- which overlaps occurred and which prioritization decisions were made. That could possibly be useful. Also if a rule is *never* used (because it's overshadowed by some other higher-priority rule that completely overlaps it) that is probably worth a warning.